src/pg_copy.c

```diff
@@ -0,0 +1,626 @@
+			Assert(relationOcc != NULL);
+			relationSuffixPos = (int)((relationOcc - lowerQuery) + strlen(relationName));			 
+
+			shardIntervalList = LookupShardIntervalList(tableId);
+			if (shardIntervalList == NIL)
+			{
+				ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+								errmsg("could not find any shards for query"),
+								errdetail("No shards exist for distributed table \"%s\".",
+										  relationName),
+								errhint("Run master_create_worker_shards to create shards "
+										"and try again.")));
+			}
+			if (!copyStatement->is_from) { /* Construct query selecting data from this relation */
+				uint64 processedCount;
+				char const* qualifiedName = quote_qualified_identifier(relation->schemaname,
+																	   relation->relname);
+				StringInfo newQuerySubstring = makeStringInfo();
+				List *queryList = NIL;
+				appendStringInfo(newQuerySubstring, "select * from %s", qualifiedName);
+				queryList = raw_parser(newQuerySubstring->data);
+				copyStatement->query = linitial(queryList);
+				copyStatement->relation = NULL;
+				/* Collecting data from shards will be done by select handler */
+				DoCopy(copyStatement, query, &processedCount);
+				return true;
+			}
+
+			/* construct pseudo predicate specifying condition for partition key */
+			partitionColumn = PartitionColumn(tableId);
+			columnOid = partitionColumn->vartype;
+			equalityExpr = MakeOpExpression(partitionColumn, BTEqualStrategyNumber);
+			rightOp = get_rightop((Expr *) equalityExpr);				 
+			Assert(IsA(rightOp, Const));
+			rightConst = (Const *) rightOp;
+			rightConst->constvalue = 0;
+			rightConst->constisnull = false;
+			rightConst->constbyval = get_typbyval(columnOid);
+			whereClauseList = list_make1(equalityExpr);
+
+			/* allocate column values and nulls arrays */
+			rel = heap_open(tableId, ExclusiveLock);
+			tupleDescriptor = RelationGetDescr(rel);
+			columnCount = tupleDescriptor->natts;
+			columnValues = palloc0(columnCount * sizeof(Datum));
+			columnNulls = palloc0(columnCount * sizeof(bool));
+
+			/*
+			 * We create a new memory context called tuple context, and read and write
+			 * each row's values within this memory context. After each read and write,
+			 * we reset the memory context. That way, we immediately release memory
+			 * allocated for each row, and don't bloat memory usage with large input
+			 * files.
+			 */
+			tupleContext = AllocSetContextCreate(CurrentMemoryContext,
+												 "COPY Row Memory Context",
+												 ALLOCSET_DEFAULT_MINSIZE,
+												 ALLOCSET_DEFAULT_INITSIZE,
+												 ALLOCSET_DEFAULT_MAXSIZE);
+				
+			/* 
+			 * Construct hash table used for shardId->Connection mapping.
+			 * We can not use connection cache from connection.c used by GteConnection because
+			 * we need to establish multiple connections with each nodes: one connection per shard
+			 */
+			shard2conn = CreateShard2ConnectionHash();
```

Comment:
I guess we typically use uncool naming without numbers :), e.g. CreateShardConnectionHash 

Could you put this behind a connection hash API, maybe move the logic into a separate file.